### PR TITLE
GitHub Actions CI

### LINKED ISSUE #7001: Look at using GitHub Actions for CI
**What's the problem this feature will solve?**
GitHub actions is generally available now and is a very strong CI option.
Documentation: https://help.github.com/en/actions/automating-your-workflow-with-github-actions/about-github-actions
Example Project: https://github.com/pypa/bandersnatch/tree/master/.github/workflows

**Describe the solution you'd like**
- Add in CI to run on GitHub actions and reach 100% parity to what we have today
- Evaluate further possibilities with it's event integration with GitHub itself.
  - E.g. Bots

### PR DESCRIPTION
**Pull request by callmecampos**
Testing Warehouse using GitHub Actions. We run all of our jobs in parallel using the native `ubuntu-latest` Docker containers provided by GitHub Actions.

- The testing job runs inside of a PostgreSQL [service container](https://docs.github.com/en/actions/configuring-and-managing-workflows/about-service-containers), specifically `postgres:10.1` instead of `postgres:latest`.
- ~Ideally we would cache Docker containers with GitHub Package Registry instead of Docker Hub (because of [faster build time when compared to other caching methods](https://dev.to/dtinth/caching-docker-builds-in-github-actions-which-approach-is-the-fastest-a-research-18ei)), however it doesn't yet seem possible to use for container actions because accessing even public packages on GitHub Package Registry requires an extra authentication step (see https://github.community/t/docker-pull-from-public-github-package-registry-fail-with-no-basic-auth-credentials-error/16358 for ongoing discussion). In case GitHub fixes this and allows a simpler integration of Actions and Package Registry, I've left the (skipped) job in [.github/workflows/main.yml](https://github.com/callmecampos/warehouse/blob/github-actions-ci/.github/workflows/main.yml#L55-L101).~
- We use GitHub Actions' native caching action to cache pip dependencies and node modules for the following jobs: `test`, `lint`, `deps`, `static-tests`, `static-pipeline`.

Resolves #7001. You can see the latest runs with runtimes at the bottom of this PR, or in the Checks tab.

~This won't yet be able to run on pypa/warehouse since the secret environment variables for Docker Hub haven't been set up in this repo, only in my fork.~ This now runs fine on pypa/warehouse.

### CODE REVIEW

**Comment by pavanagrawal123** [comment]
hey @callmecampos, this looks really good! I had #7881 open for this, but I didn't have much time recently to fix it up. This is a way better approach then I had going. I'm going to close #7881 in favor of this!

File: tests/conftest.py [hunk 1]
```diff
@@ -144,7 +145,13 @@ def database(request):
...
         janitor.drop()
 
-    return "postgresql://{}@{}:{}/{}".format(pg_user, pg_host, pg_port, pg_db)
+    pass_field = ""
+    if pg_pass:
+        pass_field = f":{pg_pass}"
+
+    return "postgresql://{}{}@{}:{}/{}".format(
+        pg_user, pass_field, pg_host, pg_port, pg_db
+    )
```
Comment (di) [inline tests/conftest.py:154]: Are these changes necessary for GitHub Actions?

File: Makefile [hunk 2]
```diff
@@ -115,8 +116,8 @@ ifneq ($(TRAVIS), false)
 	# We're on Travis, so we can lint static files locally
 	./node_modules/.bin/eslint 'warehouse/static/js/**' '**.js' 'tests/frontend/**' --ignore-pattern 'warehouse/static/js/vendor/**'
 	./node_modules/.bin/sass-lint --verbose
-else
-	# We're not on Travis, so we should lint static files inside the static container
+else ifneq ($(ACTIONS), true)
+	# We're not on Travis or GitHub Actions, so we should lint static files inside the static container
```
Comment (di) [inline Makefile:120]: We should do the same thing as if we were on Travis here. As-is, this is not running any static linting on Actions.

File: .github/workflows/main.yml [hunk 3]
```diff
@@ -0,0 +1,273 @@
...
+        ports:
+          # Maps tcp port 5432 on service container to the host
+          - 5432:5432
+    steps:
+      - name: Check out repository
+        uses: actions/checkout@v2
+
+      - uses: actions/setup-python@v2
+        with:
+          python-version: 3.8
```
Comment (di) [inline .github/workflows/main.yml:29]: Can we use `3.8.2` here?

File: .github/workflows/main.yml [hunk 4]
```diff
@@ -0,0 +1,273 @@
...
+
+      - name: Cache pip dependencies
+        uses: actions/cache@v2
+        env:
+          cache-name: cache-pip-deps-warehouse-p
```
Comment (di) [inline .github/workflows/main.yml:151]: What's with the `-p` on the end of this?

File: .github/workflows/main.yml [hunk 5]
```diff
@@ -0,0 +1,273 @@
...
+  static-tests:
+    runs-on: ubuntu-latest
+    steps:
+      - name: Check out repository
+        uses: actions/checkout@v2
+
+      - name: Cache node modules
+        uses: actions/cache@v2
+        env:
+          cache-name: cache-node-modules-static-dh
```
Comment (di) [inline .github/workflows/main.yml:214]: Same question here and below about the `-dh`

File: .github/workflows/main.yml [hunk 6]
```diff
@@ -0,0 +1,273 @@
+name: CI
+
+on: [push, pull_request]
+
+env:
+  CACHE_SEED: 1
```
Comment (di) [inline .github/workflows/main.yml:6]: If this is always `1`, why are we including it in the keys for the cache?

File: .github/workflows/main.yml [hunk 7]
```diff
@@ -0,0 +1,273 @@
...
+        run: sudo apt -y install libcurl4-openssl-dev libssl-dev pkg-config
+
+      - name: Install pip requirements
+        run: |
+          mkdir ./.cache && mkdir ./.cache/pip
+          chown -R $(whoami) ./.cache/pip
+          pip install -U pip setuptools wheel --cache-dir ./.cache/pip
+          pip install -r requirements/main.txt --cache-dir ./.cache/pip
+          pip install -r requirements/tests.txt --cache-dir ./.cache/pip
+          pip install -r requirements/deploy.txt --cache-dir ./.cache/pip
```
Comment (di) [inline .github/workflows/main.yml:84]: Is there a reason we can't use the default (`~/.cache/pip`) here?

File: .github/workflows/main.yml [hunk 8]
```diff
@@ -0,0 +1,273 @@
...
+          
+      - name: Setup node
+        run: |
+          curl -o- https://raw.githubusercontent.com/creationix/nvm/v0.33.0/install.sh | bash
+          export NVM_DIR="$HOME/.nvm" 
+          [ -s "$NVM_DIR/nvm.sh" ] && \. "$NVM_DIR/nvm.sh"
+          [ -s "$NVM_DIR/bash_completion" ] && \. "$NVM_DIR/bash_completion"
+          nvm install $NODE_VERSION
```
Comment (di) [inline .github/workflows/main.yml:92]: Can we just use https://github.com/actions/setup-node to set up a specific node version here instead?

File: .github/workflows/main.yml [hunk 9]
```diff
@@ -0,0 +1,273 @@
...
+          npm i -g npm
+          npm install -g gulp-cli
+          npm ci
+
+      - name: Run tests
+        run: |
+          python -m coverage run -m pytest --strict --postgresql-host localhost
+          python -m coverage html
+          python -m coverage report -m --fail-under 100
```
Comment (di) [inline .github/workflows/main.yml:101]: Why can't we just run `bin/tests` here instead?

File: .github/workflows/main.yml [hunk 10]
```diff
@@ -0,0 +1,273 @@
...
+        env:
+          # The hostname used to communicate with the PostgreSQL service container
+          POSTGRES_HOST: localhost
+          # The default PostgreSQL port
+          POSTGRES_PORT: 5432
```
Comment (di) [inline .github/workflows/main.yml:106]: Are these used? The test command above is specifying `localhost` not `POSTGRES_HOST`. Also, are these necessary if they're the defaults?

File: .github/workflows/main.yml [hunk 11]
```diff
@@ -0,0 +1,273 @@
...
+          key: ${{ runner.os }}-build-${{ env.CACHE_SEED }}-${{ github.job }}-${{ env.cache-name }}-${{ hashFiles('**/package-lock.json') }}
+          restore-keys: |
+            ${{ runner.os }}-build-${{ env.CACHE_SEED }}-${{ github.job }}-${{ env.cache-name }}-
+            ${{ runner.os }}-build-${{ env.CACHE_SEED }}-${{ github.job }}-
+            ${{ runner.os }}-build-${{ env.CACHE_SEED }}-
+            ${{ runner.os }}-build-
+            ${{ runner.os }}-
+
+      - name: Dependency Setup
+        run: npm install -D babel
```
Comment (di) [inline .github/workflows/main.yml:228]: Why not `npm ci`? Why does `babel` have to be specifically installed here?

File: .github/workflows/main.yml [hunk 12]
```diff
@@ -0,0 +1,273 @@
...
+          cache-name: cache-nvm-pipeline-dh
+        with:
+          path: ~/.nvm/versions/
+          key: ${{ runner.os }}-build-${{ env.CACHE_SEED }}-${{ github.job }}-${{ env.cache-name }}-${{ hashFiles('**/package-lock.json') }}
+          restore-keys: |
+            ${{ runner.os }}-build-${{ env.CACHE_SEED }}-${{ github.job }}-${{ env.cache-name }}-
+            ${{ runner.os }}-build-${{ env.CACHE_SEED }}-${{ github.job }}-
+            ${{ runner.os }}-build-${{ env.CACHE_SEED }}-
+            ${{ runner.os }}-build-
+            ${{ runner.os }}-
```
Comment (di) [inline .github/workflows/main.yml:251]: I think this is unnecessary if we use https://github.com/actions/setup-node instead.

File: .github/workflows/main.yml [hunk 13]
```diff
@@ -0,0 +1,273 @@
...
+
+      - name: Dependency Setup
+        run: npm install -D babel
```
Comment (di) [inline .github/workflows/main.yml:270]: Same question here about babel.

File: .github/workflows/main.yml [hunk 14]
```diff
@@ -0,0 +1,273 @@
...
+
+      - name: Static Pipeline Check
+        run: bin/static_pipeline # this periodically fails for no reason, just rerun the job if needed
```
Comment (di) [inline .github/workflows/main.yml:273]:
```suggestion
        run: bin/static_pipeline
```

File: .github/workflows/main.yml [hunk 15]
```diff
@@ -0,0 +1,273 @@
...
+          restore-keys: |
+            ${{ runner.os }}-${{ env.CACHE_SEED }}-${{ github.job }}-${{ env.cache-name }}-
+            ${{ runner.os }}-${{ env.CACHE_SEED }}-${{ github.job }}-
+            ${{ runner.os }}-${{ env.CACHE_SEED }}-
+            ${{ runner.os }}-
+
+      - name: Set cache permissions
+        run: |
+          mkdir ~/.cache && mkdir ~/.cache/pip
+          chown -R $(whoami) ~/.cache/pip
```
Comment (di) [inline .github/workflows/main.yml:164]: Is this necessary? The later pip-install commands in `make deps` don't use this custom cache directory.

File: tests/conftest.py (hunk 1 above)
Comment (callmecampos) [reply tests/conftest.py:154]: Nope, good catch. That was for when we were using the `postgres:latest` container which requires a password to access the db.

File: .github/workflows/main.yml (hunk 3 above)
Comment (callmecampos) [reply .github/workflows/main.yml:29]: Should be able to, I'll try it out.

File: .github/workflows/main.yml (hunk 4 above)
Comment (callmecampos) [reply .github/workflows/main.yml:151]: TLDR: It shouldn't be there.

The caching action is strange since it doesn't update the cache after the first time it caches dependencies, you have to change the key if you want a new one. So, I was sort of mindlessly adding characters whenever I would change stuff lol. However, I just found that it's recommended to [hash the dependency files](https://docs.github.com/en/actions/configuring-and-managing-workflows/caching-dependencies-to-speed-up-workflows#using-contexts-to-create-cache-keys) such that the keys change when the dependencies change, so I'm gonna update the .yml file to reflect that.

File: .github/workflows/main.yml (hunk 5 above)
Comment (callmecampos) [reply .github/workflows/main.yml:214]: Same as [the comment above](https://github.com/pypa/warehouse/pull/8255#discussion_r468752304). TLDR: shouldn't be there.

File: .github/workflows/main.yml (hunk 6 above)
Comment (callmecampos) [reply .github/workflows/main.yml:6]: Same as [the comment above](https://github.com/pypa/warehouse/pull/8255#discussion_r468752304). TLDR: shouldn't be there. I believe I was using it just as a generic way to update the cache.

File: .github/workflows/main.yml (hunk 7 above)
Comment (callmecampos) [reply .github/workflows/main.yml:84]: I'll test this out with the default `~/.cache/pip` now that we're using the regular `ubuntu-latest` container, but I was having issues in the multi-stage Docker build where the permissions on `~/.cache/pip` were set such that I wasn't able to read or write to it.

File: .github/workflows/main.yml (hunk 4 above)
Comment (callmecampos) [reply .github/workflows/main.yml:151]: Actually, it seems I'm already hashing the dependency files where I need to be. I think I did this because I was changing which directories were being cached (e.g. `~/.cache/pip` vs `./.cache/pip`, `~/.npm` vs `~/.npm && **/node_modules`) and needed the updated cache to reflect that. Still, I'll go ahead and change the `cache-name` to be something readable.

File: .github/workflows/main.yml (hunk 10 above)
Comment (callmecampos) [reply .github/workflows/main.yml:106]: Yep, not necessary, was using this in a previous iteration when I was trying different hosts/ports. Good catch.

**Comment by webknjaz** [comment]
@callmecampos Most (if not all) of the jobs here can be collapsed into a nice matrix. Are there any reasons for not implementing this?

File: .github/workflows/main.yml [hunk 16]
```diff
@@ -0,0 +1,272 @@
+name: CI
+
+on: [push, pull_request]
```
Comment (webknjaz) [inline .github/workflows/main.yml:3]: I think adding a cron schedule here would be nice too

File: .github/workflows/main.yml [hunk 17]
```diff
@@ -0,0 +1,272 @@
...
+
+env:
+  PYTHON_VERSION: 3.8.2
+  NODE_VERSION: 14.4.0
```
Comment (webknjaz) [inline .github/workflows/main.yml:7]: This won't be necessary if you make a matrix...

**Comment by di** [comment]
Hey @webknjaz, @callmecampos was working on this as part of his internship which is now finished. Would you like to take over here?

**Comment by webknjaz** [comment]
@di ah, I didn't know. If he is not going to keep working on this PR, I'll try to find some time myself, but no promises.
